src/i_winmusic.c

```diff
@@ -176,93 +176,115 @@ static void StreamOut(void)
+}
 
-static DWORD WINAPI PlayerProc(void)
+static void ResetDevice(void)
 {
-    HANDLE events[2] = { hBufferReturnEvent, hExitEvent };
+    int i;
 
-    while (1)
+    for (i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
     {
-        switch (WaitForMultipleObjects(2, events, FALSE, INFINITE))
-        {
-            case WAIT_OBJECT_0:
-                FillBuffer();
-                StreamOut();
-                break;
+        // RPN sequence to adjust pitch bend range (RPN value 0x0000)
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x00);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x00);
 
-            case WAIT_OBJECT_0 + 1:
-                return 0;
-        }
-    }
-    return 0;
-}
+        // reset pitch bend range to central tuning +/- 2 semitones and 0 cents
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x06, 0x02);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x26, 0x00);
 
-static void FreeSong(void)
-{
-    if (song.native_events)
-    {
-        free(song.native_events);
-        song.native_events = NULL;
+        // end of RPN sequence
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x7f);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x7f);
+
+        // reset all controllers
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x79, 0x00);
+
+        // reset pan to 64 (center)
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x0a, 0x40);
+
+        // reset reverb to 40 and other effect controllers to 0
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5b, winmm_reverb_level); // reverb
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5c, 0x00); // tremolo
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5d, winmm_chorus_level); // chorus
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5e, 0x00); // detune
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5f, 0x00); // phaser
     }
-    song.num_events = 0;
-    song.position = 0;
-    song.looping = false;
 }
 
-// Convert a multi-track MIDI file to an array of Windows MIDIEVENT structures.
+static byte gm_system_on[] = {0xf0, 0x7e, 0x7f, 0x09, 0x01, 0xf7};
+static byte master_vol[] = {0xf0, 0x7f, 0x7f, 0x04, 0x01, 0x7f, 0x7f, 0xf7};
 
-static void MIDItoStream(midi_file_t *file)
+static void FillBuffer(void)
 {
-    int i;
+    int num_events = 0;
+
+    buffer.position = 0;
 
-    int non_meta_events = 0;
+    if (initial_playback)
+    {
+        initial_playback = false;
 
-    int num_tracks =  MIDI_NumTracks(file);
-    win_midi_track_t *tracks = malloc(num_tracks * sizeof(win_midi_track_t));
+        // Send the GM System On SysEx message.
+        SendLongMsg(gm_system_on, sizeof(gm_system_on));
 
-    int current_time = 0;
+        // Send the full master volume SysEx message.
+        SendLongMsg(master_vol, sizeof(master_vol));
 
-    for (i = 0; i < num_tracks; ++i)
-    {
-        tracks[i].iter = MIDI_IterateTrack(file, i);
-        tracks[i].absolute_time = 0;
+        ResetDevice();
```

Comment:
The best order to do a reset seems to be:
1. Stop sound with "all notes off" and "all sound off." This prevents sound glitches when doing the next steps.
2. Send an essential set of reset messages to ensure even the most troublesome devices will be in a clean state.
3. Send a SysEx reset. "GM System On" is safe but some devices may choose to ignore it, hence (2).
4. Apply any user settings, such as custom reverb and chorus.

I suggest doing something like this:
```
if (initial_playback)
{
    initial_playback = false;
    AllNotesOff();
    AllSoundOff();
    ResetDevice();
    SendLongMsg(gm_system_on, sizeof(gm_system_on));
    ResetReverbChorus();
}
```
Where:
```
static void AllNotesOff(void)
{
    for (int i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
    {
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x7B, 0x00); // all notes off
    }
}

static void AllSoundOff(void)
{
    for (int i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
    {
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x78, 0x00); // all sound off
    }
}

static void ResetDevice(void)
{
    for (int i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
    {
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x79, 0x00); // reset all ctrls
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x00); // pb sens RPN LSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x00); // pb sens RPN MSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x06, 0x02); // data entry MSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x26, 0x00); // data entry LSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x7F); // null RPN LSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x7F); // null RPN MSB
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x07, 0x64); // channel volume
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x0A, 0x40); // pan
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x00, 0x00); // bank select msb
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x20, 0x00); // bank select lsb
        SendShortMsg(MIDI_EVENT_PROGRAM_CHANGE | i, 0x00, 0x00); // prog change
    }
}

static void ResetReverbChorus(void)
{
    for (int i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
    {
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5B, winmm_reverb_level); // reverb
        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5D, winmm_chorus_level); // chorus
    }
}
```

Reply:
Also, `AllNotesOff()` should be run at the end of each song loop to prevent hanging notes (doom2.wad MAP14 D_DDTBL2 is good example).

Reply:
This issue has been fixed in `mus2mid.c` https://github.com/chocolate-doom/chocolate-doom/commit/9e9bb23d201c68a63690f96ba0c73110f4d999f4 This issue seems to be rare, Choco/Crispy has no problems with this fix.

Reply:
Thanks for the link!